target-chains/ethereum/examples/oracle_swap/contract/src/OracleSwap.sol

```diff
@@ -0,0 +1,142 @@
+// price to reflect an unbalanced pool) or depositing / withdrawing funds. When deployed, the contract needs to be sent
+// some quantity of both the base and quote token in order to function properly (using the ERC20 transfer function to
+// the contract's address).
+contract OracleSwap {
+    event Transfer(address from, address to, uint amountUsd, uint amountWei);
+
+    IPyth pyth;
+
+    bytes32 baseTokenPriceId;
+    bytes32 quoteTokenPriceId;
+
+    ERC20 public baseToken;
+    ERC20 public quoteToken;
+
+    constructor(
+        address _pyth,
+        bytes32 _baseTokenPriceId,
+        bytes32 _quoteTokenPriceId,
+        address _baseToken,
+        address _quoteToken
+    ) {
+        pyth = IPyth(_pyth);
+        baseTokenPriceId = _baseTokenPriceId;
+        quoteTokenPriceId = _quoteTokenPriceId;
+        baseToken = ERC20(_baseToken);
+        quoteToken = ERC20(_quoteToken);
+    }
+
+    // Buy or sell a quantity of the base token. `size` represents the quantity of the base token with the same number
+    // of decimals as expected by its ERC-20 implementation. If `isBuy` is true, the contract will send the caller
+    // `size` base tokens; if false, `size` base tokens will be transferred from the caller to the contract. Some
+    // number of quote tokens will be transferred in the opposite direction; the exact number will be determined by
+    // the current pyth price. `pythUpdateData` is the binary pyth price update data (retrieved from Pyth's price
+    // service); this data should contain a price update for both the base and quote price feeds.
+    // See the frontend code for an example of how to retrieve this data and pass it to this function.
+    function swap(
+        bool isBuy,
+        uint size,
+        bytes[] calldata pythUpdateData
+    ) external payable {
+        uint updateFee = pyth.getUpdateFee(pythUpdateData.length);
+        pyth.updatePriceFeeds{value: updateFee}(pythUpdateData);
+
+        PythStructs.Price memory currentBasePrice = pyth.getPrice(
+            baseTokenPriceId
+        );
+        PythStructs.Price memory currentQuotePrice = pyth.getPrice(
+            quoteTokenPriceId
+        );
+
+        // Note: this code does all arithmetic with 18 decimal points. This approach should be fine for most
+        // price feeds, which typically have ~8 decimals. You can check the exponent on the price feed to ensure
+        // this doesn't lose precision.
+        uint256 basePrice = convertToUint(currentBasePrice, 18);
+        uint256 quotePrice = convertToUint(currentQuotePrice, 18);
+
+        // This computation loses precision. The infinite-precision result is between [quoteSize, quoteSize + 1]
+        // We need to round this result in favor of the contract.
+        uint256 quoteSize = (size * basePrice) / quotePrice;
+
+        if (isBuy) {
+            // (Round up)
+            quoteSize += 1;
```

Comment:
This is only true if the remainder of `size * basePrice % quotePrice` is positive, right? I guess its fine.